pkg/lifecycle/service.go

```diff
@@ -173,18 +241,26 @@ func (s *Service) stopForceful(ctx context.Context, pl *Instance) error {
 	return nil
 }
 
-// StopAll will ask all the pipelines to stop gracefully
+// StopAll will ask all the running pipelines to stop gracefully
 // (i.e. that existing messages get processed but not new messages get produced).
 func (s *Service) StopAll(ctx context.Context, reason error) {
-	for _, pl := range s.instances {
-		if pl.GetStatus() != StatusRunning && pl.GetStatus() != StatusRecovering {
+	s.m.Lock()
+	instances := s.runningPipelines
+	s.m.Unlock()
+
+	for _, rp := range instances {
+		p := rp.pipeline
+		if p.GetStatus() != pipeline.StatusRunning && p.GetStatus() != pipeline.StatusRecovering {
+			s.m.Lock()
+			delete(s.runningPipelines, p.ID)
+			s.m.Unlock()
```

Comment:
Wait, why are we deleting the pipeline here? 🤔 This should happen in the cleanup goroutine.

Reply:
Oh, shoot, you're right. My thinking was that, since we're iterating over the running pipelines, if it's stopped, then, it shouldn't be part of the running pipelines at all, but I didn't realize that this could also be part of a different goroutine that maybe still need be cleaned up.